### PR TITLE
Fix Issue #2306 / infinite recursion in ns_pool_for_zone

### PR DESCRIPTION
**Pull request by marcus0x62**
As described in [Issue #2306](https://github.com/hickory-dns/hickory-dns/issues/2306), it is possible to cause a Hickory server crash under certain conditions:

   1) There are no cached NS pools for a given zone with valid name servers
   2) Querying the parent DNS servers for a given zone only return child NS records. E.g.,
      querying IN NS for example.com only returns something like:

        example.com. IN NS ns1.example.com.
        example.com. IN NS ns2.example.com.

      with no non-child NS records. E.g.,

       example.com. IN NS ns1.someotherdomain.net.

   3) No glue records are returned with the NS records.

This will cause an infinite loop where ns_pool_for_zone calls resolve to get A/AAAA records for the NS records returned by the parent server, which since there is no pool for example.com will in turn call ns_pool_for_zone again, which triggers another call to resolve, and so on until the stack is exhausted.

This can be seen, as described in the issue if you configure BIND to act as a dummy root server for a hosted zone. It will only return glue records if those are present in its cache -- not the hosted zone.  So, if you send the BIND server an A query directly for the delegated nameserver prior to sending the recursive query through Hickory, this bug will not be triggered. If you empty caches on both servers and send a query for the hosted zone to Hickory first, the bug will be triggered.

This patch works by changing the logic in ns_pool_for_zone to only call resolve from within ns_pool_for_zone for non-child NS servers.  If, after querying for any non-child NS servers there still are not any NS servers to use for a zone, then queries for child NS servers will be sent to the parent zone of the zone being queried, but these queries are done with NameServerPool.lookup so as to avoid the possibility of infinite recursion.  The initial check for glue records is unchanged, so the overall order of priority within ns_pool_for_zone for identifying name servers to use for a pool is:

	1. Use glue records from the query to the parent zone
	2. Use resolve to query non-child nameservers returned without glue from the parent zone
	3. Use NameServerPool.lookup to try to directly resolve child nameserver records returned without glue from the parent zone nameserver.

### CODE REVIEW

**Comment by justahero** [comment]
I think you should be able to remove the `#[ignore]` attribute from the conformance test [`single_node_dns_graph_with_bind_as_peer`](https://github.com/hickory-dns/hickory-dns/blob/main/conformance/packages/conformance-tests/src/resolver/dnssec/regression.rs#L101) now, the test that exposed the behaviour. `hickory` should pass this test without getting stuck in a loop.

**Comment by marcus0x62** [comment]
> I think you should be able to remove the `#[ignore]` attribute from the conformance test [`single_node_dns_graph_with_bind_as_peer`](https://github.com/hickory-dns/hickory-dns/blob/main/conformance/packages/conformance-tests/src/resolver/dnssec/regression.rs#L101) now, the test that exposed the behaviour. `hickory` should pass this test without getting stuck in a loop.

This is done now.

File: crates/recursor/src/recursor_dns_handle.rs
Comment (japaric) [inline crates/recursor/src/recursor_dns_handle.rs:428]: nit: there's a pretty similar loop in line 356 that would be nice to refactor to avoid repetition

**Review by japaric — COMMENTED** [review]
looks reasonable to me but I'm not super familiar with this pre-existing part of the codebase

File: crates/recursor/src/recursor_dns_handle.rs
Comment (marcus0x62) [reply crates/recursor/src/recursor_dns_handle.rs:428]: That previous loop resolves the futures for the queries for non-child nameservers.  I think the situation where a parent NS returns an in-domain/child NS delegation without glue should be treated as an exception, and we should only attempt to resolve those name servers if there are no other usable records present.  That does require a bit of repetition, though.

File: crates/recursor/src/recursor_dns_handle.rs
Comment (bluejekyll) [inline crates/recursor/src/recursor_dns_handle.rs:347]: can you describe this in a little more detail? I think you mean use an `ANY` request, but I think `ANY` is being deprecated by many resolvers: https://blog.cloudflare.com/what-happened-next-the-deprecation-of-any

File: crates/recursor/src/recursor_dns_handle.rs
Comment (bluejekyll) [inline crates/recursor/src/recursor_dns_handle.rs:353]: This was probably missed in a previous review, I think the signature of `is_subzone` is incorrect, and it would be good for us to fix this so these clones aren't necessary.

Could you as part of this change change `is_subzone` from `fn is_subzone(parent: Name, child: Name)` to `fn is_subzone(parent: &Name, child: &Name)`? I don't see any reason why that shouldn't be valid, and would remove these clones.

File: crates/recursor/src/recursor_dns_handle.rs
Comment (bluejekyll) [inline crates/recursor/src/recursor_dns_handle.rs:398]: nit, other debug lines start with lowercase.

File: crates/recursor/src/recursor_dns_handle.rs
Comment (bluejekyll) [inline crates/recursor/src/recursor_dns_handle.rs:351]: These resolutions for `A` and `AAAA` are a bit repetitive. I think the original were a bit simpler, but with the added filters, I wonder if we should have a lambda or a generic function over the `RecordType` so that we can get a little more DRY? I'm just  a little concerned that if a bug is discovered here that we'll end up needing to change it across the 4 separate calls.

File: crates/recursor/src/recursor_dns_handle.rs
Comment (bluejekyll) [reply crates/recursor/src/recursor_dns_handle.rs:428]: @marcus0x62 , not sure if you both are talking past each other. But I think the logic here is identical, but the the debug log lines are different. If you think the added debug info is important, it should be fine to pass a const string into a function that performs this conversion and still adds the additional context to the debug info.

**Review by bluejekyll — CHANGES_REQUESTED** [review]

File: crates/recursor/src/recursor_dns_handle.rs
Comment (bluejekyll) [reply crates/recursor/src/recursor_dns_handle.rs:353]: looks like @djc fixed this in #2334

File: crates/recursor/src/recursor_dns_handle.rs
Comment (marcus0x62) [reply crates/recursor/src/recursor_dns_handle.rs:347]: I assume you mean the 'TODO' comment - that was already there. I didn't write it.  My understanding of it is that the current logic essentially builds a partial NS lookup pool.  If glue for at least one record is returned by the parent NS, we don't attempt to resolve any other NS records that were included in the delegation response.  I believe the idea here would be to have a config option like 'resolve_all_ns_delegations' in the recursor config to attempt to build a config group based on all provided NS records.

File: crates/recursor/src/recursor_dns_handle.rs
Comment (marcus0x62) [reply crates/recursor/src/recursor_dns_handle.rs:353]: I believe @djc put in a separate PR for this a few days ago. I can move that change into this PR if you prefer.

File: crates/recursor/src/recursor_dns_handle.rs
Comment (bluejekyll) [reply crates/recursor/src/recursor_dns_handle.rs:347]: ah, could you just remove the TODO? I don't think it's valid any more. Thanks for that clarification, I missed that it was already there.

File: crates/recursor/src/recursor_dns_handle.rs
Comment (marcus0x62) [reply crates/recursor/src/recursor_dns_handle.rs:428]: I'm not really worried about preserving the debug lines, but rather not firing off additional NS queries that may not be necessary and that are likely to end in failure (and prolonging failure in the process.)  I think the case where a parent NS has glue but doesn't return it is pretty unusual, although [RFC 1034](https://datatracker.ietf.org/doc/rfc1034/) leaves the door open to that as an implementation choice.  [RFC 9471](https://datatracker.ietf.org/doc/rfc9471/) (Section 3.1) says resolvers

> "MUST include all available glue records for in-domain name servers in the additional section or MUST set TC=1 if constrained by message size."

File: crates/recursor/src/recursor_dns_handle.rs
Comment (marcus0x62) [reply crates/recursor/src/recursor_dns_handle.rs:347]: But, in any case, since I didn't resolve it, I left the comment in place.

Comment (marcus0x62) [reply crates/recursor/src/recursor_dns_handle.rs:347]: Sure, no problem

File: crates/recursor/src/recursor_dns_handle.rs
Comment (bluejekyll) [reply crates/recursor/src/recursor_dns_handle.rs:428]: are we talking about the same thing? these loops look identical:

```rust
             while !a_resolves.is_empty() {
                let (next, _, rest) = select_all(a_resolves).await;
                a_resolves = rest;
                match next {
                    Ok(response) => {
                        debug!("A or AAAA response: {:?}", response);
                        let ips = response.iter().filter_map(RData::ip_addr);
                        for ip in ips {
                            let udp =
                                NameServerConfig::new(SocketAddr::from((ip, 53)), Protocol::Udp);
                            let tcp =
                                NameServerConfig::new(SocketAddr::from((ip, 53)), Protocol::Tcp);
                            config_group.push(udp);
                            config_group.push(tcp);
                        }
                    }
                    Err(e) => {
                        warn!("resolve failed {}", e);
                    }
                }
            }
```

And while we're discussing this, I think we're setting the `NameServerConfig` a little differently here on line 326:

```rust
udp.trust_negative_responses = true;
tcp.trust_negative_responses = true;
```

Do we need that in all of these NameServerConfigs?

File: crates/recursor/src/recursor_dns_handle.rs
Comment (bluejekyll) [reply crates/recursor/src/recursor_dns_handle.rs:347]: thank you.

File: crates/recursor/src/recursor_dns_handle.rs
Comment (marcus0x62) [reply crates/recursor/src/recursor_dns_handle.rs:428]: I think maybe we are.  Here's how I see the old code:

There are two distinct, sequential code blocks used to add entries to the NS pool:

1) An unconditionally fired loop on returned glue records [here](https://github.com/hickory-dns/hickory-dns/blob/e8850b788b5d10279998feb4b74ce3ffbcd1b9d6/crates/recursor/src/recursor_dns_handle.rs#L286)
2) A conditionally fired block that iterates over any NS servers that did not have glue returned [here](https://github.com/hickory-dns/hickory-dns/blob/e8850b788b5d10279998feb4b74ce3ffbcd1b9d6/crates/recursor/src/recursor_dns_handle.rs#L343).  It only runs if a) there are unresolved nameserver addresses from the query to parent NS and if the glue list is empty.

The problem with the second loop is that we attempt to resolve a child or "in-domain" nameserver address that was returned without glue, we'll get stuck in an infinite loop.

The new code has three distinct, sequential blocks:

1) An unconditionally fired loop on returned glue records [here](https://github.com/marcus0x62/hickory-dns/blob/845f3e4666e0246ab553b7ee62eb2064c8c3902c/crates/recursor/src/recursor_dns_handle.rs#L287). This loop has not changed.
2) A conditionally fired block on any non-child nameservers [here](https://github.com/marcus0x62/hickory-dns/blob/845f3e4666e0246ab553b7ee62eb2064c8c3902c/crates/recursor/src/recursor_dns_handle.rs#L348) In addition to the original checks (that the config group is empty and that there are records in the need_ips_for_names vec,) this block filters out child nameservers from the delegation set and does not attempt to resolve them.
3. A conditionally fired block on any child nameservers [here](https://github.com/marcus0x62/hickory-dns/blob/845f3e4666e0246ab553b7ee62eb2064c8c3902c/crates/recursor/src/recursor_dns_handle.rs#L397) that does the same thing as block 2 on child-nameservers, but only if there still aren't any entries in the config group and which doesn't recursively call resolve (to avoid the infinite recursion problem.)

So, I guess ultimately, I don't have a problem moving the inner loops [here](https://github.com/marcus0x62/hickory-dns/blob/845f3e4666e0246ab553b7ee62eb2064c8c3902c/crates/recursor/src/recursor_dns_handle.rs#L370) and [here](https://github.com/marcus0x62/hickory-dns/blob/845f3e4666e0246ab553b7ee62eb2064c8c3902c/crates/recursor/src/recursor_dns_handle.rs#L428) used to resolve the lookup futures into a shared function or something, but I don't think it would be a good idea to fold steps two and three *overall* into a single block.

Comment (bluejekyll) [reply crates/recursor/src/recursor_dns_handle.rs:428]: > So, I guess ultimately, I don't have a problem moving the inner loops [here](https://github.com/marcus0x62/hickory-dns/blob/845f3e4666e0246ab553b7ee62eb2064c8c3902c/crates/recursor/src/recursor_dns_handle.rs#L370) and [here](https://github.com/marcus0x62/hickory-dns/blob/845f3e4666e0246ab553b7ee62eb2064c8c3902c/crates/recursor/src/recursor_dns_handle.rs#L428) used to resolve the lookup futures into a shared function or something, 

yes, that's what I was referring to, not sure if @japaric was thinking of something beyond that.

> but I don't think it would be a good idea to fold steps two and three overall into a single block.

agreed.

> I can't think of a reason why we would want to only trust negative responses from nameservers that we learned the addresses of via glue records, which is what the current behavior effectively is.

Agreed. I think we either don't "trust" any, or we use the same setup for each.

Comment (marcus0x62) [reply crates/recursor/src/recursor_dns_handle.rs:408]: @bluejekyll I moved what I could from the inner loop into a shared function in [crates/recursor/src/lib.rs](https://github.com/hickory-dns/hickory-dns/blob/ba2862bd88ca9259fc65ed1736852918929409fe/crates/recursor/src/lib.rs#L140), but that ended up just being the innermost iteration over the ip range.  The rest of the code had enough type differences (caused by the differing return types in resolve vs lookup) to prevent an easy collapse into a single function (at least that I could figure out.)

Doing it this way did allow me to fold the config group calls from the glue IP section in as well.

File: crates/recursor/src/recursor_dns_handle.rs
Comment (marcus0x62) [reply crates/recursor/src/recursor_dns_handle.rs]: I simplified this a bit; please take a look and let me know what you think.  The remaining duplication is down the differences in nameserver selection (child vs non-child NS) and using resolve in the non-child NS section and lookup in the child section and the resulting return type differences.

File: crates/recursor/src/recursor_dns_handle.rs
Comment (bluejekyll) [reply crates/recursor/src/recursor_dns_handle.rs]: @japaric, do you have any open concerns? otherwise, I think we can close out this conversation.

File: crates/recursor/src/recursor_dns_handle.rs
Comment (bluejekyll) [reply crates/recursor/src/recursor_dns_handle.rs]: Yes, I like that simplification, thank you!

**Review by bluejekyll — APPROVED** [review]
This looks good to me know. I think we should have @japaric provide a final review.

**Comment by djc** [comment]
@marcus0x62 I just discussed with @bluejekyll and we're very happy with the contributions you've made so far. Would you be interested in joining up as a maintainer, helping out with some of the code reviews?

File: crates/recursor/src/lib.rs
Comment (djc) [inline crates/recursor/src/lib.rs:140]: Why not make this a method on `NameServerConfigGroup`?

File: crates/recursor/src/recursor_dns_handle.rs
Comment (djc) [inline crates/recursor/src/recursor_dns_handle.rs:325]: FWIW, I prefer turbofishing over type annotations for cases like these: `.collect::<Vec<_>>()`

File: crates/recursor/src/recursor_dns_handle.rs
Comment (djc) [inline crates/recursor/src/recursor_dns_handle.rs:381]: Style nit: invert this condition and return early?

File: crates/recursor/src/recursor_dns_handle.rs
Comment (djc) [inline crates/recursor/src/recursor_dns_handle.rs:324]: Could this be `r.ip_addr()` instead?

File: crates/recursor/src/lib.rs
Comment (marcus0x62) [reply crates/recursor/src/lib.rs]: I moved this to NameServerConfigGroup::append_ips

File: crates/recursor/src/recursor_dns_handle.rs
Comment (marcus0x62) [reply crates/recursor/src/recursor_dns_handle.rs:377]: I looked at this - it would result in some duplication in constructing the GenericNameServerPool.

**Comment by marcus0x62** [comment]
@djc I made most of the changes you suggested, except for the early return just due to the duplication it would cause.  Please take a look and let me know if you have any additional feedback.

After moving add_config_group_recs to be a method of NameServerConfigGroup I also refactored the config_group related calls in ns_pool_for_referral to use the new method to reduce duplication a bit more.

@japaric please let me know if you have any feedback on this. I think @bluejekyll was waiting on your input on this.

File: crates/recursor/src/recursor_dns_handle.rs
Comment (marcus0x62) [reply crates/recursor/src/recursor_dns_handle.rs]: I made this change along with the others you suggested. Most were straightforward. FuturesUnordered ended up being just a bit simpler than using select, but while making the change I also went through and cleaned up most of the remaining repetition in ns_pool_for_zone and ns_pool_for_referral.

 Practically speaking, I moved the FuturesUnordered loop into a macro which takes a closure as one of its arguments to process the lookup data from the futures. This allowed for dealing with the type variants (with both the futures themselves and the results) without a lot of overhead.

**Review by djc — COMMENTED** [review]
Some feedback left. See also my PR (please squash it into your change):

https://github.com/marcus0x62/hickory-dns/pull/1

File: crates/recursor/src/recursor_dns_handle.rs
Comment (japaric) [inline crates/recursor/src/recursor_dns_handle.rs:45]: nit: at a glance I would think this macro can be written as a generic function but if that ends up being too complex, this macro is fine.

File: crates/resolver/src/config.rs
Comment (japaric) [reply crates/resolver/src/config.rs:805]: this is down to personal preference but I generally prefer `for` loops over the `for_each` iterator adapter

**Review by japaric — COMMENTED** [review]
thanks, code-wise this looks good to me though I can't comment on the changes in the domain-specific logic (y'all know this part of DNS better than I do).

I have no opinions in terms of cleaning up the commit history.

File: crates/recursor/src/recursor_dns_handle.rs
Comment (djc) [reply crates/recursor/src/recursor_dns_handle.rs:45]: (Did that in https://github.com/marcus0x62/hickory-dns/pull/1.)

File: crates/resolver/src/config.rs
Comment (marcus0x62) [reply crates/resolver/src/config.rs:800]: this is done now.

**Comment by marcus0x62** [comment]
> Some feedback left. See also my PR (please squash it into your change):
> 
> [marcus0x62#1](https://github.com/marcus0x62/hickory-dns/pull/1)

Thanks for the generic function

File: crates/resolver/src/config.rs
Comment (djc) [inline crates/resolver/src/config.rs:805]: Nit: this really belongs in the other commit.

File: crates/recursor/src/recursor_dns_handle.rs
Comment (djc) [inline crates/recursor/src/recursor_dns_handle.rs:326]: I'm a little hesitant to actually ask for this since we've already done so many iterations, but: I think this commit could benefit from being split up again, with first the change to introduce `append_ips_from_lookup()` (and converting all the existing callers to that), and only then introducing the actual fix. That would make it a lot easier to see the semantic changes being introduced here.

Only if you're up for it, I'll approve this as is.

**Review by djc — APPROVED** [review]

File: crates/recursor/src/recursor_dns_handle.rs
Comment (marcus0x62) [reply crates/recursor/src/recursor_dns_handle.rs:326]: No problem, I split the recursion fix into a different commit.

File: crates/resolver/src/config.rs
Comment (marcus0x62) [reply crates/resolver/src/config.rs:805]: done

File: crates/recursor/src/recursor_dns_handle.rs
Comment (djc) [reply crates/recursor/src/recursor_dns_handle.rs:326]: Awesome!

**Pull request merged**
